### ISSUE TITLE
Make tsc runnable on Nashorn

### ISSUE THREAD

**Issue by markelliot** [Microsoft/TypeScript#1789]
Would be useful to be able to use gradle+jvm to run tsc directly to reduce complexity in our build system (today we use gradle to run node to run grunt to run tsc).

Some recent improvements in the JDK suggest this is likely now a fast enough environment ([1]). Others have tried in the past, so it is definitely possible ([2],[3]).

[1]:https://blogs.oracle.com/nashorn/entry/nashorn_performance_work_in_the
[2]:http://softwarecenturion.me/posts/2014-04-07-jdk8-nashorn-performance/
[3]:https://github.com/coudy/typescript/compare/9ac01de...1.0-nashorn

**Comment by bavardage** [comment]
:+1:

**Comment by RyanCavanaugh** [comment]
Looks like someone's done the basic work here: https://github.com/coudy/typescript . We can discuss whether or not taking a PR on this would be appropriate.

**Comment by DanielRosenwasser** [comment]
First, before anyone tries to work with that directly, that Nashorn system implementation is for the old compiler. Still should be easy enough to rewrite.

Second, note the following from [the somewhat recent Nashorn announcement](https://blogs.oracle.com/nashorn/entry/nashorn_performance_work_in_the):

> These plans will also solidify more for future Java releases, and we hope that Nashorn eventually will turn into a kind of "LLVM for dynamic languages on the JVM". The optimistic typing problem comes back to every dynamic language. **To prove this, our excellent thesis student Andreas Gabrielsson has successfully implemented almost everything required to run TypeScript on Nashorn.** At the time of this writing, modules in the only large feature missing. Nashorn goes directly to bytecode from TypeScript without generating JavaScript in a mid-level compiler tier. This is pretty awesome and shows how extensible Nashorn already is. We have thought about Ruby and Groovy running in a similar environment. Nashorn TypeScript is still not open sourced, but we hope we have the chance to do it soon.

Not that this should dissuade contributions, but something we should consider discussing further.

**Comment by markelliot** [comment]
My suspicion is that even when TS code is directly runnable on Nashorn that we would still need a Nashorn-specific IO module for tsc, so that tsc itself were also runnable.

**Comment by basarat** [comment]
@markelliot I suspect you will *inevitably* need node for *some* other project that does something useful with you JavaScript/CSS

**Comment by DanielRosenwasser** [comment]
> Nashorn-specific IO module for tsc

Or a Node-compatible interface exposed from Java.

**Comment by DanielRosenwasser** [comment]
> Or [a Node-compatible interface exposed from Java](https://blogs.oracle.com/java-platform-group/entry/node_js_and_io_js).

The linked blog post mentions:
* https://github.com/nodyn/nodyn
* https://github.com/apigee/trireme

**Comment by wmono** [comment]
I was able to run tsc on Avatar.js without much trouble, though I have not tested it thoroughly. You might be interested in a proof-of-concept Maven plugin that I put together this week: [Maven TypeScript Compiler Plugin](https://github.com/wmono/tsc-maven-plugin)

**Comment by mhegazy** [comment]
thanks @wmono for sharing. You mention in your description you mention
> What about watching the directory and recompiling when a file changes?

Actually tsc on node supports --watch mode that allows for that. just wanted to make sure you know about it.

**Comment by wmono** [comment]
Thanks @mhegazy I've added a tsc:watch goal to run with that option and it does seem to work on Avatar.js.

**Comment by martyglaubitz** [comment]
That would enable [Typescript Gradle Plugin](https://github.com/sothmann/typescript-gradle-plugin) to run the compiler without having nodejs installed on the system.

**Comment by Vadorequest** [comment]
:+1: I also need to use Nashorn instead of Node to run TypeScript, in Gradle as well. Does the situation has changed since the Nashorn announcement?

**Comment by martyglaubitz** [comment]
@basarat not necessarily - alot of people (especially on the enterprise) are using JS for their frontend stuff only. Yet we'd like to use Typescript in our stack - which in turn requires the compiler to be a bit more agnostic about the environment in which he runs.

**Comment by ctaggart** [comment]
This would be great to see! It would be nice to get contacts at Oracle to see if they still [plan on open sourcing](https://blogs.oracle.com/nashorn/entry/nashorn_performance_work_in_the) their Nashorn TypeScript. Otherwise, Apigee's [rowboat](https://github.com/apigee/rowboat) project is set to use Nashorn instead of Rhino for the JavaScript Engine like [trireme](https://github.com/apigee/trireme) does. It may be worth trying out.

**Comment by angelozerr** [comment]
Just for your information, I have implemented [ternjs server](http://ternjs.net/) to consume it in Java context with [tern.java](https://github.com/angelozerr/tern.java) with several JavaScript engine (node.js, [J2V8](https://github.com/eclipsesource/J2V8), Nashorm and Rhino) and the performance is not very good with Nashorn (I have not benchamrk, but when I consume ternjs with Nashorm (like open completion with a lot of JS files), Eclipse freezes a little although with node.js it is very fast)

In other words with ternjs, for performance I have : node.js > J2V8 > Nashorm > Rhino.

**Comment by ctaggart** [comment]
Thanks @angelozerr! Looking at the J2V8 docs, it looks similar to [Edge.js](https://github.com/tjanczuk/edge), but for Java instead of .NET. I think that will work.

**Comment by vojtechhabarta** [comment]
I usually run TypeScript compiler using `node`. When I have Java project with TypeScript code and I don't want to force others to install Node.js I use `frontend-maven-plugin` which installs `node` and `npm` automatically.

But I also wanted to allow users to customize my Java application using TypeScript expressions, snippets, plugins... For this feature I need to run TypeScript compiler inside Java VM so I created PR.

With this PR it should be possible to implement for Java something like [typescript-simple](https://github.com/teppeis/typescript-simple).

**Comment by vladima** [comment]
@vojtechhabarta  for embedded scripting can you just load `typescriptServices.js` and then use [ts.transpileModule](https://github.com/Microsoft/TypeScript/wiki/Using-the-Compiler-API#transpiling-single-file) to convert .ts to .js?

**Comment by vojtechhabarta** [comment]
@vladima I was thinking that TypeScript services also need `System` that's why I wanted first make `tsc` work. But apparently `ts.transpileModule` doesn't use `System`.

`ts.transpileModule` is probably sufficient in my case. I just tried to compile "Hello world" program with `typescriptServices.js` running on Nashorn and it worked. It took about 10 seconds which is about 4 times faster than `tsc` running on Nashorn. This might be usable in cases where caching is possible (and ts code doesn't change often).

Thanks

**Comment by mhegazy** [comment]
Given the perf characteristics of running the compiler on nashorn (see more in https://github.com/Microsoft/TypeScript/pull/8565#issuecomment-221334914), and the support requirements on the core ts team, we will have to decline this request in the time being. we can revisit this in the future if any of these assumptions change.

**Comment by wizawu** [comment]
https://1c.const.one/

The purpose of this project is not just to _Make tsc runnable on Nashorn_. Just wanna make JDK and Maven libraries really programmable in TypeScript (without quirky hacking).
